### DISCUSSION TITLE
Masonry like grid with ui.table

### DISCUSSION THREAD

**Discussion by h-quer** [zauberzeug/nicegui#2095] [Q&A] [answered]
### Question

The Quasar documentation has a nice example of a masonry like grid: https://quasar.dev/vue-components/table#example--masonry-like-grid

I have, so far without success, been trying to get this to work with NiceGUI. The grid style table display works in general by simply calling `table.props(add='grid')` but the customisation doesn't work. I think the relevant part is:

```html
<div class="q-pa-xs col-xs-12 col-sm-6 col-md-4">
  <q-card flat bordered>
    <q-card-section class="text-center">
      Calories for
      <br>
      <strong>{{ props.row.name }}</strong>
    </q-card-section>
    <q-separator />
    <q-card-section class="flex flex-center" :style="{ fontSize:  (props.row.calories / 2) + 'px' }">
      <div>{{ props.row.calories }} g</div>
    </q-card-section>
  </q-card>
</div>
```

I've tried adding a props slot via `table.add_slot('props', """the code above""")`, which doesn't seem to do anything. I've tried using `table.props(add="""the code above""")` but that works only partially: The card layout changes to flat, but the card content is still the default (simply all rows of my table). It seems to ignore the card section parts.

How would I go about making this simple example work within NiceGUI? Once I've got that working, I think I'll be able to build from there.
In the discussion [here](https://github.com/zauberzeug/nicegui/discussions/2002) it was mentioned that not all Qusar functionalities are implemented? Is it possible at all?

Thanks!

---

For those interested in a bit of background on why I'm trying to get this going, this is what I'm working on: https://github.com/zauberzeug/nicegui/discussions/2094

The cards layout shown in the screenshots there does look reasonably nice and is working, but performance for large lists is not ideal and using the grid style table gives access to awesome table functions like pagination and search. I'd very much like to use those. With the masonry like grid working, I think it should be possible to adapt this to essentially keep the same cards layout I have now, but inside the table, simply by adjusting the q-card definition from the Quasar example. If there is an easier way to reach that goal, I'm of course happy for suggestions!

**Comment by falkoschindler** [comment, marked as answer]
Here's a proof of concept:
```py
columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name'},
    {'name': 'age', 'label': 'Age', 'field': 'age'},
]
rows = [
    {'name': 'Alice', 'age': 18},
    {'name': 'Bob', 'age': 21},
    {'name': 'Carol', 'age': 42},
]
table = ui.table(columns=columns, rows=rows, row_key='name').props('grid')
table.add_slot('item', r'''
  <q-card flat bordered :props="props" class="m-1">
    <q-card-section class="text-center">
      <strong>{{ props.row.name }}</strong>
    </q-card-section>
    <q-separator />
    <q-card-section class="text-center">
      <div>{{ props.row.age }} years</div>
    </q-card-section>
  </q-card>
''')
```
<img width="295" alt="Screenshot 2023-11-29 at 21 18 28" src="https://github.com/zauberzeug/nicegui/assets/5767091/3480dbfc-74f7-4ebc-b525-3d3797e29c2c">

So you basically add the "grid" prop to the table and define a template for the "item" slot.

**Comment by h-quer** [comment]
Right, I read the `v-slot:item="props"` as needing to use the props slot, but in hindsight it is absolutely obvious that it would translate to using the item slot ... thanks a lot! It's working now, at least in principle, and from here it's just a matter of making "pretty" cards as far as I can tell.

Maybe the example you posted might even be a good one to add to the documentation? Card styles are quite popular.

**Reply by falkoschindler** [reply]
@h-quer Good idea. I added the demo in commit https://github.com/zauberzeug/nicegui/commit/cbc083d310c1b5b319f9de8dd5ed334d2676d1b3.
